centreon/src/Core/ResourceAccess/Application/UseCase/UpdateRule/UpdateRule.php

```diff
@@ -0,0 +1,529 @@
+        } catch (RuleException $exception) {
+            $presenter->presentResponse(
+                match ($exception->getCode()) {
+                    RuleException::CODE_CONFLICT => new ConflictResponse($exception),
+                    default => new ErrorResponse($exception),
+                }
+            );
+            $this->error($exception->getMessage(), ['trace' => $exception->getTraceAsString()]);
+        } catch (\Throwable $exception) {
+            $presenter->presentResponse(
+                new ErrorResponse(RuleException::updateRule())
+            );
+            $this->error((string) $exception);
+        }
+    }
+
+    /**
+     * @param Rule $rule
+     * @param UpdateRuleRequest $request
+     *
+     * @throws \Throwable
+     */
+    private function updateInTransaction(Rule $rule, UpdateRuleRequest $request): void
+    {
+        try {
+            $this->debug('Starting resource access rule update transaction process');
+            $this->dataStorageEngine->startTransaction();
+            $this->updateBasicInformation($rule, $request);
+
+            $this->validator->assertContactsAndContactGroupsAreNotEmpty(
+                $request->contactIds,
+                $request->contactGroupIds
+            );
+            $this->updateLinkedContacts($rule, $request);
+            $this->updateLinkedContactGroups($rule, $request);
+            $this->updateResourceLinks($request);
+
+            $this->debug('Commit resource access rule update transaction process');
+            $this->dataStorageEngine->commitTransaction();
+        } catch (\Throwable $exception) {
+            $this->error("Rollback of 'Update resource access rule' transaction");
+            $this->dataStorageEngine->rollbackTransaction();
+
+            throw $exception;
+        }
+    }
+
+    /**
+     * This method will ensure that the ids provided for the dataset filters are valid
+     * and that the structure of the dataset is correct regarding the hierarchy defined in the DatasetFilter entity.
+     *
+     * @param UpdateRuleRequest $request
+     *
+     * @throws RuleException
+     * @throws \InvalidArgumentException
+     *
+     * @return DatasetFilter[]
+     */
+    private function validateAndCreateDatasetFiltersFromRequest(UpdateRuleRequest $request): array
+    {
+        $datasetFilters = [];
+
+        $validateAndBuildDatasetFilter = function (
+            array $data,
+            ?DatasetFilter $parentDatasetFilter
+        ) use (&$validateAndBuildDatasetFilter, &$datasetFilter): void {
+            /**
+             * In any case we want to make sure that
+             *     - resources provided are valid (exist)
+             *     - the datasetfilter type provided is valid (validated by entity)
+             *     - that the dataset filter hierarchy is valid (validated by entity).
+             */
+            $this->validator->assertResourceIdsAreValid($data['type'], $data['resources']);
```

Comment:
To simplify reading, you can remove 'assertResourceIdsAreValid' from this closure and put it just after line 231 to separate assert and build.